bot/cogs/watchchannels/watchchannel.py

```diff
@@ -0,0 +1,359 @@
+import asyncio
+import datetime
+import logging
+import re
+from abc import ABC, abstractmethod
+from collections import defaultdict, deque
+from typing import Optional
+
+import aiohttp
+import discord
+from discord import Color, Embed, Message, Object, errors
+from discord.ext.commands import BadArgument, Bot, Context
+
+from bot.constants import BigBrother as BigBrotherConfig, Guild as GuildConfig
+from bot.pagination import LinePaginator
+from bot.utils import messages
+from bot.utils.time import time_since
+
+log = logging.getLogger(__name__)
+
+URL_RE = re.compile(r"(https?://[^\s]+)")
+
+
+def proxy_user(user_id: str) -> Object:
+    try:
+        user_id = int(user_id)
+    except ValueError:
+        raise BadArgument
+    user = Object(user_id)
+    user.mention = user.id
+    user.display_name = f"<@{user.id}>"
+    user.avatar_url_as = lambda static_format: None
+    user.bot = False
+    return user
+
+
+class WatchChannel(ABC):
+    """
+    Base class for WatchChannels
+
+    Abstracts the basic functionality for watchchannels in
+    a granular manner to allow for easy overwritting of
+    methods in the child class.
+    """
+
+    @abstractmethod
+    def __init__(self, bot: Bot, destination, webhook_id, api_endpoint, api_default_params, logger) -> None:
+        """
+        abstractmethod for __init__ which should still be called with super().
+
+        Note: Some of the attributes below need to be overwritten in the
+        __init__ of the child after the super().__init__(*args, **kwargs)
+        call.
+        """
+        self.bot = bot
+
+        self.destination = destination  # E.g., Channels.big_brother_logs
+        self.webhook_id = webhook_id  # E.g.,  Webhooks.big_brother
+        self.api_endpoint = api_endpoint  # E.g., 'bot/infractions'
+        self.api_default_params = api_default_params  # E.g., {'active': 'true', 'type': 'watch'}
+        self.log = logger  # Logger of the child cog for a correct name in the logs
+
+        # These attributes can be left as they are in the child class
+        self._consume_task = None
+        self.watched_users = defaultdict(dict)
+        self.message_queue = defaultdict(lambda: defaultdict(deque))
+        self.consumption_queue = {}
+        self.retries = 5
+        self.retry_delay = 10
+        self.channel = None
+        self.webhook = None
+        self.message_history = [None, None, 0]
```

Comment:
Yeah, I had this on my "I might change this later" list, but I didn't get around to it. I'll leave this conversation open for now and implement it later. Maybe we can discuss a good approach in the core-dev channel first. A dataclass could work and has nice attribute access, like named tuples. Nothing wrong with adding a dataclass to a module that implements an ABC either, I think.

Reply:
I've created a `dataclass` for this for now. I think it enhances readability, but it may be a bit overkill. I'll leave this convo open as a reminder for now.

Reply:
The dataclass could be simplified by removing unpacking support. While mildly convenient, it's really only used in one place. I think it would make the if statement that uses the message history too long, but I wonder what the net line count will be considering the `__iter__` uses up 3 lines. 

That being said I am not opposed to unpacking support remaining.

Reply:
I've removed the `__iter__` and turned this into a multiline `if`-statement of the sort @lemonsaurus hates. If he wants to murder me to death over it, I'll change it to the assigned condition version he prefers.